### ISSUE TITLE
[Question]: Multi Agents on Single Host with Docker in Docker

### ISSUE THREAD

**Issue by skinnyskinner** [microsoft/azure-pipelines-agent#4251]
### Describe your question

Hello, after a lot of debugging an trying I ran into an issue when running docker in docker on linux.

So I currently want to run multiple agents on one host maschine, just to use up more performance from older hardware.
I want to run everything in docker on ubuntu server 22.04, so I started creating a docker agent that can also start other containers, where the job is executed.

After some time I was able to start the job with the right directories (since the mount is hard to change)

Now to the question/issue, when a job gets started, it cleans all the other containers with the label f3b821
with this line in the initizalize containers

```
docker ps --all --quiet --no-trunc --filter "label=f3b821"
```

After that all containers are cleaned, that are currently running on the socket with the label. (all containers are created with the label)

This is an issue, since the agents use the same docker socket, in order to save downloads on the internal network, otherwise i get a load of 12 agents pulling a 4GB image all at once.

Is there a proper way to allow the same socket and multiple agents on the same machine? Like override the label per job or agent?

Current Docker-Compose for agent starts:

```yaml
version: '3'

volumes:
  agent-1-azp:
    name: agent-1-azp
  agent-2-azp:
    name: agent-2-azp

services:
  azure-agent-1:
    image: azureagent
    build: .
    env_file:
      - ./.env
    privileged: true
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - $PWD/agent-1-work:$PWD/agent-1-work
      - agent-1-azp:/azp
    environment:
      - AZP_AGENT_NAME=agent-1
      - AZP_WORK=$PWD/agent-1-work

  azure-agent-2:
    image: azureagent
    env_file:
      - ./.env
    privileged: true
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - $PWD/agent-2-work:$PWD/agent-2-work
      - agent-2-azp:/azp
    environment:
      - AZP_AGENT_NAME=agent-2
      - AZP_WORK=$PWD/agent-2-work
```
In the .env file are only secrets with AZP_URL, AZP_TOKEN, AZP_POOL

### Versions

Agent 3.218.0 on/in ubuntu:20.04

### Environment type (Please select at least one enviroment where you face this issue)

- [ ] Self-Hosted
- [ ] Microsoft Hosted
- [ ] VMSS Pool
- [X] Container

### Azure DevOps Server type

dev.azure.com (formerly visualstudio.com)

### Operation system

ubuntu:20.04

**Comment by skinnyskinner** [comment]
Looks like my Question is a big Challenge.
Are there any news/update what I can do in this case?

**Comment by skinnyskinner** [comment]
This issue still persists, strange that there is absolutely no one working at microsoft.

**Comment by skinnyskinner** [comment]
Is there any news about the triage?

**Comment by skinnyskinner** [comment]
I found a solution to my problem.
After looking into the code I figured out the hash `f3b821` gets created by the Directory, no idea why a directory should be the unique identifier for something like this, but ok.
Since i have the agents in a docker container the hash is always the same.

My solution now was to change the directory in the start.sh script so the hash is different.
In my case i used the agentName for the directory and added this:
```
if [[ ! -e $AZP_AGENT_NAME ]]; then
  mkdir -p $AZP_AGENT_NAME
fi

cd $AZP_AGENT_NAME
```

Now i can have multiple agents on the same docker socket.
